distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso

```diff
@@ -1,14 +1,182 @@
+import project.Any.Any
+import project.Data.Numbers.Decimal
 import project.Data.Numbers.Integer
+import project.Data.Numbers.Number
 import project.Error.Common.Type_Error
 import project.Error.Error
+import project.Error.Incomparable_Values.Incomparable_Values
+import project.Error.Unimplemented.Unimplemented
+import project.Nothing
+import project.Meta
+import project.Meta.Atom
+from project.Data.Boolean import all
+
+## Provides custom ordering, equality check and hash code for types that need it.
+
+    The Enso runtime system offers default implementation of _equality_
+    as well as capability to _compute hash code_ (for use in `Map`) automatically.
+    The default implementation is sufficient for most of the programming activities.
+    Especially when defining new type and its constructors, they get sensible
+    implementation of both functions.
+
+    Should there be a need to redefine the default implementation, here is a way:
+    Define conversion function `Comparable.from` for your `type` and return pointer to
+    another `type` that satisfies either of the following two definitions:
+
+    ```
+    type Ordered_Comparator T
+        is_ordered = True
+        compare : T -> T -> Ordering
+        hash : T -> Integer
+
+    type Unordered_Comparator T
+        is_ordered = False
+        equals : T -> T -> Boolean
+        hash : T -> Integer
+    ```
+
+    Or `Incomparable` in case that the type `T` should not be compared at all.
+
+    Note that there has to be `is_ordered` method defined which returns a Boolean
+    indicating that the comparator is ordered. This is currently needed as there is
+    no way how to define interfaces in Enso.
+
+    An _unordered comparator_ has to implement both `equals` and `hash` to define
+    a _total_ custom equality. By _total_, we mean that every instance of the type
+    has to be either equal or not equal, which is represented by the type signature
+    of `equals` - just `Boolean` is returned without any errors thrown.
+
+    An _ordered comparator_ has `compare` method instead of `equals` method, that is
+    expected to return `Ordering` signaling that an instance of the type is either
+    less than, equal to, or greater than the other instance. This relation is also
+    _total_, meaning that all the instances of the type are comparable.
+
+    The runtime expects the following semantics for all the comparators:
+    - Hash consistency:
+      - If x == y then hash(x) == hash(y)
+      - If hash(x) != hash(y) then x != y
+    - Consistency: if x == y then x == y for all the subsequent invocations.
+    - Symmetry: if x == y then y == x
```

Comment:
Symmetry could be enforced. Just like we check `hash` code consistency we could also verify (from time to time) that `x == y` and `y == x` is consistent.